File: packages/python/topographic-adjustment-core/src/btm_topography/weights.py
"""STAR*NET-compatible observation standard-error propagation."""

from __future__ import annotations

import math

from .geometry import ARCSEC_TO_RAD


def effective_sigmas(
    *,
    slope_distance_m: float,
    zenith_rad: float,
    direction_arcsec: float,
    zenith_arcsec: float,
    distance_mm: float,
    distance_ppm: float,
    instrument_centering_m: float,
    target_centering_m: float,
    vertical_centering_m: float,
    edm_std_error_model: str = "additive",
) -> dict[str, float]:
    """Propagate default measurement and centering errors per STAR*NET manual.

    The direction/zenith centering terms are angular (radians).  Distance components remain
    in metres. STAR*NET combines distance constant and ppm additively by default; the
    root-sum-square mode is used only when `.EDM PROPAGATE` is explicitly selected.
    """

    numeric_inputs = {
        "slope_distance_m": slope_distance_m,
        "zenith_rad": zenith_rad,
        "direction_arcsec": direction_arcsec,
        "zenith_arcsec": zenith_arcsec,
        "distance_mm": distance_mm,
        "distance_ppm": distance_ppm,
        "instrument_centering_m": instrument_centering_m,
        "target_centering_m": target_centering_m,
        "vertical_centering_m": vertical_centering_m,
    }
    if any(not math.isfinite(value) for value in numeric_inputs.values()):
        raise ValueError("Weight inputs must be finite")
    if slope_distance_m <= 0:
        raise ValueError("Slope distance must be greater than zero")
    if edm_std_error_model not in {"additive", "propagated"}:
        raise ValueError("edm_std_error_model must be additive or propagated")
    for name, value in numeric_inputs.items():
        if name not in {"slope_distance_m", "zenith_rad"} and value < 0:
            raise ValueError(f"{name} must be non-negative")
    horizontal = abs(slope_distance_m * math.sin(zenith_rad))
    vertical = slope_distance_m * math.cos(zenith_rad)
    horizontal = max(horizontal, 1e-9)
    centering2 = instrument_centering_m**2 + target_centering_m**2
    sigma_hz = math.sqrt((direction_arcsec * ARCSEC_TO_RAD) ** 2 + centering2 / horizontal**2)
    constant_m = distance_mm / 1000.0
    proportional_m = slope_distance_m * distance_ppm * 1e-6
    sigma_sd_measurement = (
        math.hypot(constant_m, proportional_m)
        if edm_std_error_model == "propagated"
        else constant_m + proportional_m
    )
    sigma_sd = math.sqrt(
        sigma_sd_measurement**2
        + (horizontal / slope_distance_m) ** 2 * centering2
        + 2.0 * (vertical / slope_distance_m) ** 2 * vertical_centering_m**2
    )
    sigma_vz = math.sqrt(
        (zenith_arcsec * ARCSEC_TO_RAD) ** 2
        + (vertical / slope_distance_m) ** 2 * centering2 / slope_distance_m**2
        + 2.0 * (horizontal / slope_distance_m) ** 2 * vertical_centering_m**2 / slope_distance_m**2
    )
    result = {"hz": sigma_hz, "vz": sigma_vz, "sd": sigma_sd}
    if any(value <= 0 for value in result.values()):
        raise ValueError("Each effective observation sigma must be greater than zero")
    return result

File: src/domain/math/weights.ts
import { ARCSEC2RAD } from '@/domain/math/geometry';

export type EdmStdErrorModel = 'additive' | 'propagated';

export interface TotalStationWeightInput {
  slopeDistanceM: number;
  zenithRad: number;
  directionArcSec: number;
  zenithArcSec: number;
  distanceMm: number;
  distancePpm: number;
  instrumentCenteringM: number;
  targetCenteringM: number;
  verticalCenteringM: number;
  edmStdErrorModel?: EdmStdErrorModel;
}

export interface EffectiveTotalStationSigmas {
  hzRad: number;
  hzArcSec: number;
  vzRad: number;
  vzArcSec: number;
  sdM: number;
  distanceMeasurementM: number;
}

function requireFinite(value: number, field: string): number {
  if (!Number.isFinite(value)) throw new Error(`${field} must be finite`);
  return value;
}

/**
 * Combines the constant and proportional EDM terms using STAR*NET's selected model.
 * STAR*NET defaults to ADDITIVE; PROPAGATED is available through `.EDM PROPAGATE`.
 */
export function distanceMeasurementSigmaM(
  slopeDistanceM: number,
  distanceMm: number,
  distancePpm: number,
  model: EdmStdErrorModel = 'additive',
): number {
  requireFinite(slopeDistanceM, 'slopeDistanceM');
  requireFinite(distanceMm, 'distanceMm');
  requireFinite(distancePpm, 'distancePpm');
  if (slopeDistanceM <= 0) throw new Error('slopeDistanceM must be greater than zero');
  if (distanceMm < 0 || distancePpm < 0) throw new Error('EDM error terms must be non-negative');
  if (model !== 'additive' && model !== 'propagated') throw new Error('EDM model must be additive or propagated');
  const constantM = distanceMm / 1000;
  const proportionalM = slopeDistanceM * distancePpm * 1e-6;
  return model === 'propagated'
    ? Math.hypot(constantM, proportionalM)
    : constantM + proportionalM;
}

/**
 * STAR*NET-compatible effective standard errors for one slope-distance/zenith sight.
 * The returned values already include instrument, target and vertical centering and can
 * therefore be emitted as explicit `DM` standard errors without `.ADDCENTERING ON`.
 */
export function effectiveTotalStationSigmas(input: TotalStationWeightInput): EffectiveTotalStationSigmas {
  const numeric = Object.entries(input).filter(([key]) => key !== 'edmStdErrorModel') as Array<[string, number]>;
  for (const [field, value] of numeric) requireFinite(value, field);
  if (input.slopeDistanceM <= 0) throw new Error('slopeDistanceM must be greater than zero');
  for (const [field, value] of numeric) {
    if (field !== 'zenithRad' && field !== 'slopeDistanceM' && value < 0) {
      throw new Error(`${field} must be non-negative`);
    }
  }

  const slope = input.slopeDistanceM;
  const horizontal = Math.max(1e-9, Math.abs(slope * Math.sin(input.zenithRad)));
  const vertical = slope * Math.cos(input.zenithRad);
  const centering2 = input.instrumentCenteringM ** 2 + input.targetCenteringM ** 2;
  const hzRad = Math.sqrt((input.directionArcSec * ARCSEC2RAD) ** 2 + centering2 / horizontal ** 2);
  const distanceMeasurementM = distanceMeasurementSigmaM(
    slope,
    input.distanceMm,
    input.distancePpm,
    input.edmStdErrorModel,
  );
  const sdM = Math.sqrt(
    distanceMeasurementM ** 2
    + (horizontal / slope) ** 2 * centering2
    + 2 * (vertical / slope) ** 2 * input.verticalCenteringM ** 2,
  );
  const vzRad = Math.sqrt(
    (input.zenithArcSec * ARCSEC2RAD) ** 2
    + (vertical / slope) ** 2 * centering2 / slope ** 2
    + 2 * (horizontal / slope) ** 2 * input.verticalCenteringM ** 2 / slope ** 2,
  );
  if (!(hzRad > 0) || !(vzRad > 0) || !(sdM > 0)) {
    throw new Error('Each effective observation sigma must be greater than zero');
  }

  return {
    hzRad,
    hzArcSec: hzRad / ARCSEC2RAD,
    vzRad,
    vzArcSec: vzRad / ARCSEC2RAD,
    sdM,
    distanceMeasurementM,
  };
}
